contracts/VRT/XVSVesting.sol

```diff
@@ -0,0 +1,257 @@
+        uint256 vestingCount = vestingsOfRecipient.length;
+        uint256 totalWithdrawableAmount = 0;
+
+        for(uint i = 0; i < vestingCount; i++) {
+            VestingRecord storage vesting = vestingsOfRecipient[i];
+            (uint256 vestedAmount, uint256 toWithdraw) = calculateWithdrawableAmount(vesting.amount, vesting.startTime, vesting.withdrawnAmount);
+            if(toWithdraw > 0){
+                totalWithdrawableAmount = totalWithdrawableAmount.add(toWithdraw);
+                vesting.withdrawnAmount = vesting.withdrawnAmount.add(toWithdraw);
+            }
+        }
+
+       if(totalWithdrawableAmount > 0){
+           uint256 xvsBalance = xvs.balanceOf(address(this));
+           require(xvsBalance >= totalWithdrawableAmount, "Insufficient XVS for withdrawal");
+           emit XVSWithdrawn(recipient, totalWithdrawableAmount);
+           xvs.safeTransfer(recipient, totalWithdrawableAmount);
+       }
+    }
+
+    function getWithdrawableAmount(address recipient) view public nonZeroAddress(recipient) 
+    returns (uint256 totalWithdrawableAmount, uint256 totalVestedAmount, uint256 totalWithdrawnAmount)
+    {
+        VestingRecord[] memory vestingsOfRecipient = vestings[recipient];
+        uint256 vestingCount = vestingsOfRecipient.length;
+
+        for(uint i = 0; i < vestingCount; i++) {
+            VestingRecord memory vesting = vestingsOfRecipient[i];
+            (uint256 vestedAmount, uint256 toWithdraw) = calculateWithdrawableAmount(vesting.amount, vesting.startTime, vesting.withdrawnAmount);
+            totalVestedAmount = totalVestedAmount.add(vestedAmount);
+            totalWithdrawableAmount = totalWithdrawableAmount.add(toWithdraw);
+            totalWithdrawnAmount = totalWithdrawnAmount.add(vesting.withdrawnAmount);
+        }
+
+        return (totalWithdrawableAmount, totalVestedAmount, totalWithdrawnAmount);
+    }
+
+    function calculateWithdrawableAmount(uint256 amount, uint256 vestingStartTime, uint256 withdrawnAmount)
+      view internal returns (uint256, uint256) {
+        uint256 vestedAmount = calculateVestedAmount(amount, vestingStartTime, getCurrentTime());
+        uint toWithdraw = vestedAmount.sub(withdrawnAmount);
+        return (vestedAmount, toWithdraw);
+    }
+
+    function getVestedAmount(address recipient) view public nonZeroAddress(recipient) returns (uint256) {
+
+        VestingRecord[] memory vestingsOfRecipient = vestings[recipient];
+        uint256 vestingCount = vestingsOfRecipient.length;
+        uint256 totalVestedAmount = 0;
+        uint256 currentTime = getCurrentTime();
+
+        for(uint i = 0; i < vestingCount; i++) {
+            VestingRecord memory vesting = vestingsOfRecipient[i];
+            uint256 vestedAmount = calculateVestedAmount(vesting.amount, vesting.startTime, currentTime);
+            totalVestedAmount = totalVestedAmount.add(vestedAmount);
+        }
+
+        return totalVestedAmount;
+    }
```

Comment:
Looks like the logic is the same as in `getWithdrawableAmount`, and it's not called from within the contract (so we don't need to worry much about the gas spendings) – do we need a separate `getVestedAmount` function then?

Reply:
@kkirka 
this is for UI/client to display amount vested for an address.
we can make it external

Reply:
@kkirka 
it includes `vestedAmount`. same math as in `getVestedAmount`. 
infact client can use the function `getWithdrawableAmount` 

`getWithdrawableAmount`  - 3 values:  `totalWithdrawableAmount` , `totalVestedAmount` , `totalWithdrawnAmount`
`getVestedAmount` is to get only vestedAmount

Reply:
@kkirka 
shall i remove the `getVestedAmount`
and let client use only `getWithdrawableAmount` and infer the value needed from response ?